### ISSUE TITLE
xgboost training continuation convergence problem

### ISSUE THREAD

**Issue by ronaldpereira** [dmlc/xgboost#5712]
Hello :)

I was using xgboost to train in a large dataset (using dmatrix external memory feature) and using the AWS Sagemaker with spot training enabled, so my training job can be interrupted at any moment. So, I needed to save xgboost.train() checkpoints. I was saving it at each round, but the model was giving a much worse score than the full trained one with just a subset of the data. So I decided to make some small tests on the iris dataset and I got to the following results:

1 - Full train:
```
xgb_full = xgb.train(hyperparameters,
                     dtrain,
                     num_boost_round=100,
                     evals=[(dtrain, 'train')],
                     feval=f1_score_dmatrix,
                     maximize=True,
                     verbose_eval=True)
```
 This one converges to `train-error:0.00833	train-f1:0.98667`

2 - Checkpoint at each round:
```
xgb_partial = None
for _ in range(100):
    xgb_partial = xgb.train(hyperparameters,
                            dtrain,
                            num_boost_round=1,
                            evals=[(dtrain, 'train')],
                            feval=f1_score_dmatrix,
                            maximize=True,
                            verbose_eval=True,
                            xgb_model=xgb_partial)
```
This one converges to `train-error:0.04167	train-f1:0.93671`

3 - Checkpoint at each 10 rounds:
```
xgb_partial = None
for _ in range(10):
    xgb_partial = xgb.train(hyperparameters,
                            dtrain,
                            num_boost_round=10,
                            evals=[(dtrain, 'train')],
                            feval=f1_score_dmatrix,
                            maximize=True,
                            verbose_eval=True,
                            xgb_model=xgb_partial)
```
This one converges to `train-error:0.01667	train-f1:0.97368`

In that way, is this a known issue for training continuation feature and/or expected behavior?

The full test script code can be found in this [gist](https://gist.github.com/ronaldpereira/5ded29087c07a7c793783d8dd0390258)

**Comment by trivialfis** [comment]
Hi, could you share which xgboost version are you using?

**Comment by trivialfis** [comment]
We have some strict tests for doing training continuation.  If you are using latest XGBoost (1.1 as of writing), I will try to reproduce it.

**Comment by ronaldpereira** [comment]
> Hi, could you share which xgboost version are you using?

xgboost 1.1.0 version.

**Comment by trivialfis** [comment]
Just reproduced the issue.  It works fine when without specifying subsample though.

**Comment by ronaldpereira** [comment]
> Just reproduced the issue. It works fine when without specifying subsample though.

Just confirmed this as well. It works as expected when removing the subsample hyperparameter. Maybe the subsampling is not the same when loading the model (e.g. different seeds)?

**Comment by hcho3** [comment]
@ronaldpereira Try setting the seed manually and see if you can make subsampling deterministic. I'm afraid that subsamping may vary between multiple invocations of `xgb.train()`, even with a `seed` parameter. The reason is that setting the seed multiple times during the training results in a different random number sequence than setting the seed only once.

My suggestion is to remove the subsampling to achieve determinism with preemptible spot workers.
